### PR TITLE
feat: auth & redirect to portal from cloud

### PR DESCRIPTION
**Pull request by gcusnieux**
## Issue

https://gravitee.atlassian.net/browse/CJ-1192

## Description

A small description of what you did in that PR.

### CODE REVIEW

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/auth/CockpitAuthenticationResourceTest.java [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+    @Test
+    public void shouldAuthenticateToConsole() {
+        UserEntity userEntity = mockUserEntity();
+
+        String jwt = this.generateJWT(userEntity.getId(), ORGANIZATION_ID, ENVIRONMENT_ID, "audience", null, null, null);
+        when(userService.findBySource(ORGANIZATION_ID, "cockpit", userEntity.getId(), true)).thenReturn(userEntity);
+        when(authoritiesProvider.retrieveAuthorities(userEntity.getId(), ORGANIZATION_ID, ENVIRONMENT_ID)).thenReturn(Set.of());
+        when(installationAccessQueryService.getConsoleUrl(ORGANIZATION_ID)).thenReturn(rootTarget("console").getUri().toString());
+        final Response response = rootTarget(PATH).queryParam("token", jwt).request().get();
+        assertEquals(HttpStatusCode.OK_200, response.getStatus());
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/auth/CockpitAuthenticationResourceTest.java]: Is it possible to check the url built in the resource? I'm surprise to see 200 status when in the code when send a redirect response that should have a 307 🤔

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/auth/CockpitAuthenticationResourceTest.java [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+        String jwt = this.generateJWT(userEntity.getId(), ORGANIZATION_ID, ENVIRONMENT_ID, "audience", null, null, "portal");
+        when(userService.findBySource(ORGANIZATION_ID, "cockpit", userEntity.getId(), true)).thenReturn(userEntity);
+        when(authoritiesProvider.retrieveAuthorities(userEntity.getId(), ORGANIZATION_ID, ENVIRONMENT_ID)).thenReturn(Set.of());
+        when(installationAccessQueryService.getPortalAPIUrl(ENVIRONMENT_ID)).thenReturn(rootTarget("portal").getUri().toString());
+        final Response response = rootTarget(PATH).queryParam("token", jwt).request().get();
+        assertEquals(HttpStatusCode.OK_200, response.getStatus());
+    }
+
+    @Test
+    public void shouldAuthenticateToPortalFromRequest() {
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/auth/CockpitAuthenticationResourceTest.java:190]: I don't understand this case compared to the previous one 🤔

**Review by jgiovaresco — COMMENTED** [review]
It could be great to have a PR description that summarizes the change... It would help the review...

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/auth/CockpitAuthenticationResourceTest.java (hunk 1 above)
Comment (gcusnieux) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/auth/CockpitAuthenticationResourceTest.java]: Yes, I was also surprised during the tests, but at that point the redirection was done and the response came from the [final resources](https://github.com/gravitee-io/gravitee-api-management/pull/7118/commits/4a153515a20d506b7849ed9b94661581dc71ad48#diff-79fec14c50e1ad22c9e2f520854cbbb0cf69af0dfad488d890a626524b2c9fbaR271-R287). 
I couldn't find a way to see the url, but we could add a header or something if necessary. WDYT ?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/auth/CockpitAuthenticationResourceTest.java (hunk 2 above)
Comment (gcusnieux) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/auth/CockpitAuthenticationResourceTest.java:190]: This is to validate [this case](https://github.com/gravitee-io/gravitee-api-management/pull/7118/files#diff-5ca355170a04d6131f478773617e694dd05a5dad5d9cacc477e002dbf2eca8fbR172-R181)
I must confess, I got it from the `PortalRedirectResource`

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/auth/CockpitAuthenticationResourceTest.java (hunk 1 above)
Comment (jgiovaresco) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/auth/CockpitAuthenticationResourceTest.java]: Could you add in that method https://github.com/gravitee-io/gravitee-api-management/blob/master/gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/JerseySpringTest.java#L119 the following line

```
config.property(ClientProperties.FOLLOW_REDIRECTS, false);
```

From what I understood it should disable redirection

**Review by jgiovaresco — COMMENTED** [review]
Can you try what I suggested https://github.com/gravitee-io/gravitee-api-management/pull/7118/files#r1551241697

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/auth/CockpitAuthenticationResourceTest.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    public void shouldAuthenticateToPortal() {
+        UserEntity userEntity = mockUserEntity();
+
+        String jwt = this.generateJWT(userEntity.getId(), ORGANIZATION_ID, ENVIRONMENT_ID, "audience", null, null, "portal");
+        when(userService.findBySource(ORGANIZATION_ID, "cockpit", userEntity.getId(), true)).thenReturn(userEntity);
+        when(authoritiesProvider.retrieveAuthorities(userEntity.getId(), ORGANIZATION_ID, ENVIRONMENT_ID)).thenReturn(Set.of());
+        when(installationAccessQueryService.getPortalAPIUrl(ENVIRONMENT_ID)).thenReturn(rootTarget("portal").getUri().toString());
+        final Response response = rootTarget(PATH).queryParam("token", jwt).request().get();
+        assertEquals(HttpStatusCode.TEMPORARY_REDIRECT_307, response.getStatus());
+        assertTrue(response.getLocation().toString().contains("/portal/environments/environment-id/auth/console?token="));
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/auth/CockpitAuthenticationResourceTest.java:185]: From what I saw, we should be able to check the token. What do you think about adding some assertions on the token to strengthen the test?

**Review by jgiovaresco — APPROVED** [review]

**Review by aurelien-pacaud — APPROVED** [review]

**Pull request merged**